### DISCUSSION TITLE
Hierarchical menu partial

### DISCUSSION THREAD

**Discussion by waseemsadiq** [joomlatools/joomlatools-pages#601] [Snippets]
This partial can an unlimited hierarchical menu.The active menu is activated using a `active-child` class, the parent and grandparents of the active menu get a `child-is-active` class.

#### Pages

The starting folder can be configured using the `$folder`variable, if not defined the root will be used  and the depth using the `$level`variable, if not defined the menu will render the complete pages tree.

```php
<? $items = collection('pages', ['folder' => $folder ?? '.', 'level' => $level ?? null,  'recurse' => 'true']) ?>
```

#### Joomla

The menu type can be configured using the `$type`variable, if not defined the root will be used  and the depth using the `$level`variable, if not defined the menu will render the complete menu items tree.

```php
<? $items = collection('menus', ['type' => $type ?? 'mainmenu', 'level' => $level ?? null,  'recurse' => 'true']) ?>
```

### The partial

```php
<? $renderMenu = function($items, $level) use(&$renderMenu) { ?>

    <? foreach($items as $item):

        $item_active = false;
        if(url()->getPath() == '/'.$item->path) :
            $item_active = true;
        endif;

        $child_active = false;
        if($item->hasChildren()) :
            ob_start();
            $child_active = $renderMenu($item->getChildren(), $level + 1);
            $menu = ob_get_clean();
        endif;

        $class = $item_active ? 'is-active' : ($child_active ? 'child-is-active' : '');
        ?>

        <li data-level="<?= $level ?>"<?= $class ? ' class="'.$class.'"' : ''?>>
            <a href="<?= route($item) ?>" itemprop="url">
                <span itemprop="name"><?= $item->name ?></span>
            </a>
            <? if($item->hasChildren()) : ?>
                <ul>
                    <?= $menu ?>
                </ul>
            <? endif ?>
        </li>

    <?
    endforeach;
    return $item_active ?? false
    ?>

<? } ?>

<ul class="nav" itemscope itemtype="http://www.schema.org/SiteNavigationElement">
    <? $renderMenu($items, 1) ; ?>
</ul>

```

**Comment by manojLondhe** [comment]
Thanks. How can we get this from Joomla's menu manager instead building it from pages?

**Reply by johanjanssens** [reply]
Not something we have done already but technically possible, would need to create a collection that can integrated with the Joomla menu items table, or even with the Joomla SiteMenu API. 

The SiteMenu object offers a getItems() method that allows to return a list of specific menu items: https://github.com/joomla/joomla-cms/blob/3.10-dev/libraries/src/Menu/SiteMenu.php Wrapping that in a collection would give you very easy access to the menu items.

If you need help creating that let le me know.

**Reply by manojLondhe** [reply]
thank you :+1:

**Reply by johanjanssens** [reply]
With the latest changes here: https://github.com/joomlatools/joomlatools-pages/pull/680 and https://github.com/joomlatools/joomlatools-pages/pull/363/commits/9e9525406de71078632b23f55042c6f65f09ab56 You can now use `routet()`for Joomla menu entities too.  Updated the partial info to reflect that.

**Comment by johanjanssens** [comment]
@waseemsadiq  I have updated the partial to support unlimited nested levels and no longer require javascript to set the `is-active` and `child-is-active`classes. The partial handles this itself now.

**Reply by waseemsadiq** [reply]
Awesome sauce sir! Much more preferable than doing it with JS 👍🏾

**Comment by manojLondhe** [comment]
Getting notice on PHP 7.4

```Unparenthesized `a ? b : c ? d : e` is deprecated. Use either `(a ? b : c) ? d : e` or `a ? b : (c ? d : e)```

change from
`$class = $item_active ? 'is-active' : $child_active ? 'child-is-active' : '';`
=>

to 
` $class = $item_active ? 'is-active' : ($child_active ? 'child-is-active' : '');`

**Reply by johanjanssens** [reply]
Good one, updated the code.

**Comment by manojLondhe** [comment]
thank you @waseemsadiq @johanjanssens for this
